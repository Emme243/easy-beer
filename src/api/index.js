import { BeersApiService } from '@/api/entities/beersApi';

export const $api = {
  beers: new BeersApiService(),
};
