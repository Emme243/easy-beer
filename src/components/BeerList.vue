<template>
  <div v-if="isLoading" class="grid-list">
    <v-skeleton-loader
      min-width="250"
      v-for="i in 6"
      :key="i"
      :boilerplate="true"
      type="image, list-item-two-line, button"
    />
  </div>
  <v-alert v-else-if="hasError" type="error">
    Ha habido un error al traer las cervezas, por qu√© no te relajas, te tomas una y lo vuelves a
    intentar m√°s tarde üòâ
  </v-alert>
  <v-alert v-else-if="beers.length === 0" type="info">
    <h3 class="text-h5">No hay cervezas que mostrar con el criterio solicitado.</h3>
    <span>Intenta reducir los criterios de b√∫squeda (barra de b√∫squeda y filtros)</span>
    <ResetAllFiltersButton color="blue darken-4" />
  </v-alert>
  <div v-else class="grid-list mb-3 mt-5">
    <BeerCard v-for="beer in beers" :key="`${beer.id}-${beer.name}`" :beer="beer" />
  </div>
</template>

<script>
import BeerCard from '@/components/Card/BeerCard';
import ResetAllFiltersButton from '@/components/FilterModal/ResetButtons/ResetAllFiltersButton';
import { mapGetters } from 'vuex';

const beersPerPage = 15;
export default {
  name: 'BeerList',
  components: { ResetAllFiltersButton, BeerCard },
  data() {
    return {
      hasError: false,
      isLoading: true,
      beers: [],
    };
  },
  methods: {
    fetchBeers() {
      console.log('fetchBeers');
      this.isLoading = true;
      this.hasError = false;
      this.$api.beers
        .getAll(this.apiQueryParams)
        .then(beers => {
          this.beers = beers || [];
        })
        .catch(() => {
          this.hasError = true;
        })
        .finally(() => {
          this.isLoading = false;
        });
    },
  },
  computed: {
    ...mapGetters({
      page: 'filter/page',
      searchQuery: 'filter/searchQuery',
      minAbvValue: 'filter/minAbvValue',
      maxAbvValue: 'filter/maxAbvValue',
      brewedAfter: 'filter/initialBrewedMonth',
      brewedBefore: 'filter/finalBrewedMonth',
      filterStore: 'filter/filterStore',
    }),
    apiQueryParams() {
      return {
        per_page: beersPerPage,
        page: this.page,
        ...(this.minAbvValue !== 0 &&
          this.maxAbvValue !== 0 && { abv_gt: this.minAbvValue, abv_lt: this.maxAbvValue }),
        ...(this.searchQuery && { beer_name: this.searchQuery }),
        ...(this.brewedAfter && {
          brewed_after: this.brewedAfter.split('-').reverse().join('-'),
        }),
        ...(this.brewedBefore && {
          brewed_before: this.brewedBefore.split('-').reverse().join('-'),
        }),
      };
    },
  },
  watch: {
    filterStore() {
      this.fetchBeers();
    },
  },
};
</script>

<style scoped lang="scss">
.grid-list {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
  grid-gap: 30px;
}
</style>
