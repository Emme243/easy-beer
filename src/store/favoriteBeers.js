export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    addBeerToFavorites({ commit }, beer) {
      commit('addBeerToFavorites', beer);
    },
    removeBeerFromFavorites({ commit }, beerId) {
      commit('removeBeerFromFavorites', beerId);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
};
