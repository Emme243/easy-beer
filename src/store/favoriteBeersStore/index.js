export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    handleBeerInFavorites({ commit, getters }, beer) {
      const isBeerInFavorites = getters.isBeerInFavorites(beer.id);
      if (isBeerInFavorites) commit('removeBeerFromFavorites', beer.id);
      else commit('addBeerToFavorites', beer);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
};
