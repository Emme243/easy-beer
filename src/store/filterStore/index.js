import { abvRangeStore } from '@/store/filterStore/abvRangeStore';
import { brewedDateStore } from '@/store/filterStore/brewedDateStore';
import { paginationStore } from '@/store/filterStore/paginationStore';
import { searchQueryStore } from '@/store/filterStore/searchQueryStore';

export const filterStore = {
  namespaced: true,
  modules: {
    abvRange: abvRangeStore,
    brewedDate: brewedDateStore,
    page: paginationStore,
    searchQuery: searchQueryStore,
  },
  state: () => ({}),
  getters: {
    filterStore(_, getters) {
      return {
        minAbvValue: getters.minAbvValue,
        maxAbvValue: getters.maxAbvValue,
        initialBrewedMonth: getters.initialBrewedMonth,
        finalBrewedMonth: getters.finalBrewedMonth,
        page: getters.page,
        searchQuery: getters.searchQuery,
      };
    },
    areFiltersWithDefaultValue(_, getters) {
      return (
        getters.isAbvRangeWithDefaultValue &&
        !getters.initialBrewedMonth &&
        !getters.finalBrewedMonth &&
        !getters.searchQuery
      );
    },
    numberOfFiltersApplied(_, getters) {
      let numberOfFiltersApplied = 0;
      if (!getters.isAbvRangeWithDefaultValue) numberOfFiltersApplied++;
      if (getters.initialBrewedMonth || getters.finalBrewedMonth) numberOfFiltersApplied++;
      return numberOfFiltersApplied;
    },
  },
  mutations: {},
  actions: {
    setAllFilters({ dispatch }, filterQuery) {
      dispatch('setPage', filterQuery.page);
      dispatch('setMinAbvValue', filterQuery.minAbvValue);
      dispatch('setMaxAbvValue', filterQuery.maxAbvValue);
      dispatch('setInitialBrewedMonth', filterQuery.initialBrewedMonth);
      dispatch('setFinalBrewedMonth', filterQuery.finalBrewedMonth);
      dispatch('setSearchQuery', filterQuery.searchQuery);
    },
    resetAllFilters({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('setSearchQuery', '');
      dispatch('resetPage');
    },
    resetFiltersInFilterModal({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('resetPage');
    },
  },
};
