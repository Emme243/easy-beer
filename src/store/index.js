import { favoriteBeersStore } from '@/store/favoriteBeersStore';
import { filterStore } from '@/store/filterStore';

import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    filter: filterStore,
    favoriteBeers: favoriteBeersStore,
  },
});
