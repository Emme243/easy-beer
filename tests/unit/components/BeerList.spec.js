import { $api } from '@/api';
import BeerList from '@/components/BeerList';
import store from '@/store';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import Vuetify from 'vuetify';
import { beersMock } from '../mocks/beersMock';

global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve(beersMock),
  })
);

describe('Pruebas en el componente beerList.vue', function () {
  let vuetify;
  const localVue = createLocalVue();

  // router mock
  const router = { push: jest.fn() };
  // route mock
  const route = { query: {} };
  const mountFunction = options => {
    return shallowMount(BeerList, {
      localVue,
      vuetify,
      store,
      mocks: { $route: route, $router: router, $api },
      ...options,
    });
  };

  beforeEach(() => {
    fetch.mockClear();
    vuetify = new Vuetify();
  });

  const beerLoaderSelector = '[data-test-id="beer-loader"]';
  const beerErrorMessageSelector = '[data-test-id="beer-error-message"]';
  const noBeerMessageSelector = '[data-test-id="no-beer-message"]';
  const beerContainerSelector = '[data-test-id="beer-container"]';

  it('Debe hacer match con el snapshot', () => {
    const wrapper = mountFunction();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Debe mostrar un loader cuando se carga la lista de cervezas', () => {
    const wrapper = mountFunction({ data: () => ({ isLoading: true }) });
    expect(wrapper.find(beerLoaderSelector).exists()).toBe(true);
  });

  it('Debe mostrar un mensaje de error cuando se produce un error al cargar la lista de cervezas', () => {
    const wrapper = mountFunction({ data: () => ({ isLoading: false, hasError: true }) });
    expect(wrapper.find(beerErrorMessageSelector).exists()).toBe(true);
  });

  it('Debe mostrar un mensaje al usuario cuando no se encuentran cervezas', () => {
    const wrapper = mountFunction({
      data: () => ({ isLoading: false, hasError: false, beers: [] }),
    });
    expect(wrapper.find(noBeerMessageSelector).exists()).toBe(true);
  });

  it('Debe mostrar la lista de cervezas', () => {
    const wrapper = mountFunction({
      data: () => ({ isLoading: false, hasError: false, beers: beersMock }),
    });
    const cards = wrapper.findAll(`${beerContainerSelector} > beerinfocard-stub`);
    expect(cards.length).toBe(beersMock.length);
    expect(wrapper.find(beerContainerSelector).exists()).toBe(true);
  });

  it('Cuando el filterStore cambia, se debe disparar el método updateQueryParamsInUrl', async () => {
    const wrapper = mountFunction();
    const spy = jest.spyOn(wrapper.vm, 'updateQueryParamsInUrl');
    store.commit('filter/setPage', 2);
    await wrapper.vm.$nextTick();
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('El filterStore no debe dispararse cuando se monte el componente', () => {
    const wrapper = mountFunction({});
    const spy = jest.spyOn(wrapper.vm, 'fetchBeers');
    expect(spy).toHaveBeenCalledTimes(0);
  });

  it('Cuando el filterStore cambiar, se debe disparar el método fetchBeers', async () => {
    const wrapper = mountFunction({});
    const spy = jest.spyOn(wrapper.vm, 'fetchBeers');
    store.commit('filter/setInitialBrewedMonth', '01-2019');
    await wrapper.vm.$nextTick();
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('Cuando varias propiedades del filterStore cambian, se debe disparar el método fetchBeers una vez', async () => {
    const wrapper = mountFunction({});
    const spy = jest.spyOn(wrapper.vm, 'fetchBeers');
    store.commit('filter/setPage', 2);
    store.commit('filter/setInitialBrewedMonth', '01-2019');
    store.commit('filter/setFinalBrewedMonth', '02-2022');
    store.commit('filter/setMinAbvValue', 7);
    await wrapper.vm.$nextTick();
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
