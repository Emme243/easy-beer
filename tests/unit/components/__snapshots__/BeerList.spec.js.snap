// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente beerList.vue Debe hacer match con el snapshot 1`] = `
<div data-test-id="beer-loader" class="grid-list">
  <v-skeleton-loader-stub minwidth="250" boilerplate="true" type="image, list-item-two-line, button" types="[object Object]"></v-skeleton-loader-stub>
  <v-skeleton-loader-stub minwidth="250" boilerplate="true" type="image, list-item-two-line, button" types="[object Object]"></v-skeleton-loader-stub>
  <v-skeleton-loader-stub minwidth="250" boilerplate="true" type="image, list-item-two-line, button" types="[object Object]"></v-skeleton-loader-stub>
  <v-skeleton-loader-stub minwidth="250" boilerplate="true" type="image, list-item-two-line, button" types="[object Object]"></v-skeleton-loader-stub>
  <v-skeleton-loader-stub minwidth="250" boilerplate="true" type="image, list-item-two-line, button" types="[object Object]"></v-skeleton-loader-stub>
  <v-skeleton-loader-stub minwidth="250" boilerplate="true" type="image, list-item-two-line, button" types="[object Object]"></v-skeleton-loader-stub>
</div>
`;
